optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.solver.thread;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ThreadUtils {
+
+    protected static final transient Logger logger = LoggerFactory.getLogger(ThreadUtils.class);
+
+    public static void shutdownAwaitOrKill(ExecutorService executor, String logIndentation, String name) {
+        // Intentionally clearing the interrupted flag so that awaitTermination() in step 3 works.
+        if (Thread.interrupted()) {
+            // 2a. If the current thread is interrupted, propagate interrupt signal to children by initiating
```

Comment:
Looks like this code was refactored into this new class from `DefaultPartitionedSearchPhase.java`. Probably should at least remove the numbers in the comments since there is no longer a step 1 otherwise its a big confusing.

Reply:
Yes, we need to fix that, but I think we might need to simplify the whole thing here. It's probably overkilled.